### ISSUE TITLE
대한아이엠 이미지 분류 - 산업인공지능개론 

### ISSUE THREAD

**Issue by BONGKUNKIM** [BONGKUNKIM/Projects#1]
import tensorflow as tf
import os
import numpy as np

# 이미지 데이터가 들어있는 폴더 경로
data_dir = "D:/image test"

# 이미지 분류 클래스 이름
class_names = ["grid", "nonwoven", "product", "roll", "winder"]

# 데이터 전처리 및 로드
def preprocess_image(image):
    image = tf.image.decode_jpeg(image, channels=3)
    image = tf.image.resize(image, [224, 224])
    image /= 255.0
    return image

def load_and_preprocess_image(path):
    image = tf.io.read_file(path)
    return preprocess_image(image)

def load_data(data_dir):
    data = []
    labels = []
    for class_index, class_name in enumerate(class_names):
        class_dir = os.path.join(data_dir, class_name)
        for image_name in os.listdir(class_dir):
            image_path = os.path.join(class_dir, image_name)
            image = load_and_preprocess_image(image_path)
            data.append(image)
            labels.append(class_index)
    return np.array(data), np.array(labels)

# 데이터 로드
data, labels = load_data(data_dir)

# 데이터 셔플
num_samples = len(data)
shuffle_indices = np.random.permutation(num_samples)
data = data[shuffle_indices]
labels = labels[shuffle_indices]

# 데이터 분할
train_ratio = 0.8
num_train_samples = int(num_samples * train_ratio)

train_data = data[:num_train_samples]
train_labels = labels[:num_train_samples]

test_data = data[num_train_samples:]
test_labels = labels[num_train_samples:]

# CNN 모델 정의
model = tf.keras.Sequential([
    tf.keras.layers.Conv2D(32, (3, 3), activation='relu', input_shape=(224, 224, 3)),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(64, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Conv2D(128, (3, 3), activation='relu'),
    tf.keras.layers.MaxPooling2D((2, 2)),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dense(len(class_names))
])

# 모델 컴파일
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# 모델 훈련
model.fit(train_data, train_labels, epochs=20)

# 모델 평가
test_loss, test_acc = model.evaluate(test_data, test_labels)
print("테스트 정확도:", test_acc)

import matplotlib.pyplot as plt

# 모델 훈련 결과 그래프 그리기
def plot_training_results(history):
    acc = history.history['accuracy']
    loss = history.history['loss']

    epochs = range(len(acc))

    plt.plot(epochs, acc, 'r', label='Training Accuracy')
    plt.title('Training Accuracy')
    plt.legend(loc=0)
    plt.figure()

    plt.plot(epochs, loss, 'r', label='Training Loss')
    plt.title('Training Loss')
    plt.legend(loc=0)
    plt.figure()
